libraries/tickfs/src/tickfs.rs

```diff
@@ -0,0 +1,930 @@
+        &self,
+        hash_function: Option<&mut H>,
+        key: Option<&[u8]>,
+        value: Option<&[u8]>,
+        buf: Option<&mut [u8]>,
+    ) -> Result<SuccessCode, ErrorCode> {
+        let ret = match self.state.get() {
+            State::AppendKey(_) => {
+                self.append_key(hash_function.unwrap(), key.unwrap(), value.unwrap())
+            }
+            State::GetKey(_) => self.get_key(hash_function.unwrap(), key.unwrap(), buf.unwrap()),
+            State::InvalidateKey(_) => self.invalidate_key(hash_function.unwrap(), key.unwrap()),
+            _ => unreachable!(),
+        };
+
+        match ret {
+            Ok(_) => self.state.set(State::None),
+            Err(e) => match e {
+                ErrorCode::ReadNotReady(_) | ErrorCode::EraseNotReady(_) => {}
+                _ => self.state.set(State::None),
+            },
+        }
+
+        ret
+    }
+
+    /// Appends the key/value pair to flash storage.
+    ///
+    /// `hash_function`: Hash function with no previous state. This is
+    ///                  usually a newly created hash.
+    /// `key`: A unhashed key. This will be hashed internally. This key
+    ///        will be used in future to retrieve or remove the `value`.
+    /// `value`: A buffer containing the data to be stored to flash.
+    ///
+    /// On success nothing will be returned.
+    /// On error a `ErrorCode` will be returned.
+    pub fn append_key(
+        &self,
+        hash_function: &mut H,
+        key: &[u8],
+        value: &[u8],
+    ) -> Result<SuccessCode, ErrorCode> {
+        let (hash, region) = self.get_hash_and_region(hash_function, key);
+
+        // Length not including check sum
+        let package_length = HEADER_LENGTH + value.len();
+        let object_length = HEADER_LENGTH + value.len() + CHECK_SUM_LEN;
+
+        if object_length > 0xFFF {
+            return Err(ErrorCode::ObjectTooLarge);
+        }
+
+        // Create the header:
+        let header = ObjectHeader::new(hash, object_length as u16);
+
+        let mut region_offset: isize = 0;
+
+        loop {
+            let new_region = match self.state.get() {
+                State::None => region as isize + region_offset,
+                State::Init(state) => {
+                    match state {
+                        InitState::AppendKeyReadRegion(reg) => reg as isize,
+                        _ => {
+                            // Get the data from that region
+                            region as isize + region_offset
+                        }
+                    }
+                }
+                State::AppendKey(key_state) => match key_state {
+                    KeyState::ReadRegion(reg) => reg as isize,
+                },
+                State::GarbageCollect(state) => match state {
+                    RubbishState::ReadRegion(reg) => reg as isize,
+                    _ => unreachable!(),
+                },
+                _ => unreachable!(),
```

Comment:
What prevents a user from calling `read_key()`, then an async operation starting, but then calling `append_key()` before calling `continue_op`? I.e., why can't the current state be `GetRead`?

Reply:
This prevents that. It doesn't support two concurrent operations.

Handling that would be very complex. What if two append_key operations called, how do we then make guarantees about ordering and not loosing data?

Only 1 operation can be in progress at a time, we require the operation to complete before a new one can start.